// @zkochan/tap-diff/dist/pdiff.js
'use strict';

var _diff = require('./diff');

module.exports = {
  diff: function diff(lhs, rhs) {
    return _diff(lhs, rhs);
  },
  addLineNumbers: function addLineNumbers(diff_) {
    // Convert diff data structure to display
    var diff = [];
    var lineNumberOfLhs = 0;
    var lineNumberOfRhs = 0;
    var isInLhs = false;
    var isInRhs = false;
    var delta = null;

    var i = 0;
    while (i < diff_.length) {
      if (i != diff_.length - 1 && diff_[i].removed && diff_[i + 1].added && diff_[i].value.endsWith("\n") && diff_[i + 1].value.endsWith("\n")) {
        // Replace lines
        var removedDelta = diff_[i];
        var addedDelta = diff_[i + 1];

        // Add removedDelta to diff
        var rlines = removedDelta.value.split("\n");
        for (var j = 0; j < rlines.length - 1; j++) {
          var line = rlines[j];
          diff.push({
            lineNumberOfLhs: lineNumberOfLhs,
            values: [{
              removed: true,
              value: line
            }]
          });
          lineNumberOfLhs += 1;
        }
        // Add addedDelta to diff
        var alines = addedDelta.value.split("\n");
        for (var _j = 0; _j < alines.length - 1; _j++) {
          var _line = alines[_j];
          diff.push({
            lineNumberOfRhs: lineNumberOfRhs,
            values: [{
              added: true,
              value: _line
            }]
          });
          lineNumberOfRhs += 1;
        }

        i += 2;
      } else {
        var delta_ = diff_[i];
        var lines = delta_.value.split('\n');

        var length = delta_.value.endsWith("\n") ? lines.length - 1 : lines.length;

        for (var _j2 = 0; _j2 < length; _j2++) {
          delta = delta || {};
          delta.values = delta.values || [];

          var _line2 = lines[_j2];

          if (!delta_.removed) {
            // delta.value is in the rhs
            isInRhs = true;
          }
          if (!delta_.added) {
            // delta.value is in the lhs
            isInLhs = true;
          }

          if (_line2.length != 0) {
            if (delta_.removed) {
              delta.values.push({
                removed: true,
                value: _line2
              });
            } else if (delta_.added) {
              delta.values.push({
                added: true,
                value: _line2
              });
            } else {
              delta.values.push({
                value: _line2
              });
            }
          }

          if (_j2 != lines.length - 1 || delta_.value.endsWith("\n")) {
            // End a line
            if (isInRhs) {
              // delta.value is in the rhs
              delta.lineNumberOfRhs = lineNumberOfRhs;
              lineNumberOfRhs += 1;
            }
            if (isInLhs) {
              // delta.value is in the lhs
              delta.lineNumberOfLhs = lineNumberOfLhs;
              lineNumberOfLhs += 1;
            }
            if (delta.values.length == 0) {
              if (isInRhs && isInLhs) {
                delta.values.push({ value: "" });
              } else if (isInRhs) {
                delta.values.push({
                  added: true,
                  value: ""
                });
              } else if (isInLhs) {
                delta.values.push({
                  removed: true,
                  value: ""
                });
              }
            }
            diff.push(delta);

            // Initialize the local variables
            isInRhs = false;
            isInLhs = false;
            delta = null;
          }
        }
        i += 1;
      }
    }

    if (delta != null) {
      // Deal with a corner case (when last character of text is not \n)
      if (isInRhs) {
        // delta.value is in the rhs
        delta.lineNumberOfRhs = lineNumberOfRhs;
        lineNumberOfRhs += 1;
      }
      if (isInLhs) {
        // delta.value is in the lhs
        delta.lineNumberOfLhs = lineNumberOfLhs;
        lineNumberOfLhs += 1;
      }

      if (delta.values.length == 0) {
        if (isInRhs && isInLhs) {
          delta.values.push({ value: "" });
        } else if (isInRhs) {
          delta.values.push({
            added: true,
            value: ""
          });
        } else if (isInLhs) {
          delta.values.push({
            removed: true,
            value: ""
          });
        }
      }

      diff.push(delta);
    }

    return diff;
  },

  extractDiff: function extractDiff(diffWithLineNumbers, line) {
    if (line < 0) return [diffWithLineNumbers];

    var result = [];
    var group = [];
    var lastIndex = null;
    for (var i = 0; i < diffWithLineNumbers.length; i++) {
      var lineDelta = diffWithLineNumbers[i];

      var isChanged = lineDelta.values.some(function (elem, index, arr) {
        return elem.added || elem.removed;
      });

      if (isChanged) {
        // Add deltas near this delta
        if (lastIndex != null && lastIndex < i - line - 1) {
          // Divide a group
          result.push(group);
          group = [];
          lastIndex = null;
        }

        // Add lines
        lastIndex = lastIndex == null ? i - line - 1 : lastIndex;
        for (var j = lastIndex + 1; j <= i + line; j++) {
          if (j >= 0 && j < diffWithLineNumbers.length) {
            group.push(diffWithLineNumbers[j]);
          }
        }

        // Update lastIndex
        lastIndex = i + line;
      }
    }
    if (group.length != 0) {
      result.push(group);
    }

    return result;
  }
};
